### PR TITLE
Convolution transpose

### PR DESCRIPTION
**Pull request by hunse**
Support the ConvTransposeInc operator added in nengo/nengo#1648.

Things might be a bit cleaner if we support `GeneralConvInc` instead of `ConvInc` and `ConvTransposeInc` (then e.g. we could have one `GeneralConvSet` operator), but some code (e.g. operator merging) breaks if we don't have build methods for each operator class (if we try to do them only for a superclass).

TODO:
- [x] Test merging ConvTransposeInc

### CODE REVIEW

**Review by tbekolay — APPROVED** [review]
This LGTM and the tests pass, but as with a lot of NengoDL stuff, I don't know this material deep enough to comment on things like efficiency and correctness. Do you want to take a quick look as well @drasmuss?

File: nengo_dl/tests/test_graph_optimizer.py [hunk 1]
```diff
@@ -1111,12 +1111,26 @@ def test_remove_reset_incs():
...
-    operators = [Reset(x), ConvInc(dummies.Signal(), dummies.Signal(), x, None)]
+    conv = nengo.Convolution(1, input_shape=(3, 3, 1))
+    operators = [Reset(x), ConvInc(dummies.Signal(), dummies.Signal(), x, conv)]
     new_operators = remove_reset_incs(operators)
     assert len(new_operators) == 1
     assert isinstance(new_operators[0], transform_builders.ConvSet)
     assert new_operators[0].Y is x
 
+    # convtransposeinc converted to convtransposeset
+    if not compat.is_dummy_type(compat.ConvTransposeInc):
```
Comment (drasmuss) [inline nengo_dl/tests/test_graph_optimizer.py:1122]: Rather than checking `is_dummy_type` here, I would check the `nengo` version. It just makes it easier when we increase the minimum nengo version to find the places in the code where we can remove backwards compatibility logic.

File: nengo_dl/compat.py [hunk 2]
```diff
@@ -14,6 +14,24 @@
...
+    def __init__(self, *arg, **kwargs):  # pragma: no cover
+        raise RuntimeError("Cannot instantiate")
+
+
+def make_dummy_type(name):
+    """Return a NoType subclass with the given name."""
+    return type(name, (NoType,), {})
+
+
+def is_dummy_type(t):
```
Comment (drasmuss) [inline nengo_dl/compat.py:30]: I think we can remove this function after my other suggestions (re: checking nengo version rather than the type).

File: nengo_dl/tests/test_transforms.py [hunk 3]
```diff
@@ -4,66 +4,67 @@
...
 
+@pytest.mark.parametrize("force_cpu", (False, True))
 @pytest.mark.parametrize("channels_last", (True, False))
-def test_merge_conv(Simulator, channels_last, seed, pytestconfig):
-    from nengo.builder.transforms import (  # pylint: disable=import-outside-toplevel
-        ConvInc,
-    )
+@pytest.mark.parametrize("transpose", (True, False))
+def test_merge_conv(Simulator, transpose, channels_last, force_cpu, seed, pytestconfig):
+    if transpose and is_dummy_type(ConvolutionTranspose):
```
Comment (drasmuss) [inline nengo_dl/tests/test_transforms.py:17]: Again here I would check the nengo version rather than `is_dummy_type`

File: nengo_dl/tests/test_transforms.py [hunk 4]
```diff
@@ -4,66 +4,67 @@
...
+        pytest.skip("Nengo version does not support ConvolutionTranspose")
+
+    is_cpu = not tf_gpu_installed or pytestconfig.getoption("--device") == "/cpu:0"
```
Comment (drasmuss) [inline nengo_dl/tests/test_transforms.py:20]: I'd keep the more general `cpu in device.lower()` approach; there are lots of valid device strings (if we wanted to be fancy we could actually build the https://www.tensorflow.org/api_docs/python/tf/DeviceSpec and extract the device from that, but that's probably overkill for a test).

File: nengo_dl/transform_builders.py [hunk 5]
```diff
@@ -89,7 +115,7 @@ def build_pre(self, signals, config):
             perm_x[1:-1] = perm_x[2:]
             perm_x[-1] = 1
 
-            self.perm_x = tf.constant(perm_x)
+            self.perm_x = perm_x
```
Comment (drasmuss) [inline nengo_dl/transform_builders.py:118]: It's better to do the `tf.constant` here in `build_pre` (as that is just called once). Otherwise the np array will be converted to a tensorflow array every time the op is applied in `build_step` (which can be multiple times, e.g. once for inference and once for training, or every time the simulation step is unrolled).

File: nengo_dl/transform_builders.py [hunk 6]
```diff
@@ -165,13 +191,38 @@ def build_step(self, signals):
...
+                output_space + [output_filters]
+                if self.conv.channels_last
+                else [output_filters] + output_space
+            )
+            if self.perm_x is not None:
+                output_shape = [output_shape[i] for i in self.perm_x]
+
+            # swap channels, because conv_transpose order is for forward weights
+            dims = self.conv.dimensions
+            filters = tf.transpose(W, perm=list(range(dims)) + [dims + 1, dims])
```
Comment (drasmuss) [inline nengo_dl/transform_builders.py:208]: Similar to above comment about the `tf.constant`, I think all of this could be precomputed in `build_pre` because it doesn't depend on `X` (except for the batch size, which can just be specified as `-1` I'd expect), which would be more efficient.

**Review by drasmuss — CHANGES_REQUESTED** [review]
Took a look through, some minor comments on efficiency/version handling below but all looks good other than that :+1:

File: nengo_dl/tests/test_graph_optimizer.py (hunk 1 above)
Comment (tbekolay) [reply nengo_dl/tests/test_graph_optimizer.py:1122]: Makes sense to me :+1:

File: nengo_dl/transform_builders.py (hunk 6 above)
Comment (tbekolay) [reply nengo_dl/transform_builders.py:208]: Replacing `X.shape[0]` with `-1` seems to work (at least for the test suite).

**Comment by tbekolay** [comment]
OK, fixed those `compat.py` checks, and the new pylint warnings. I also looked into the other test failures and ended up being a NengoCore bug recently introduced. Fix that is here: https://github.com/nengo/nengo/pull/1678 and those tests should pass once that's merged in.

**Review by drasmuss — APPROVED** [review]
Fixups LGTM :+1:

**Comment by tbekolay** [comment]
Ok, pushed again, d8619d3dd5fa77cdf33abd867101dc9fc7449694 is the original version of the PR and 84e8dfa40326aa4dba416dd606a632cff3650a7e is my fixup with the `batch_size` change. You can inspect `X` right before the `tf.nn.conv_transpose` call to see what's up with that.

**Comment by tbekolay** [comment]
Squashed it all, and added [one change](https://github.com/nengo/nengo-dl/pull/183/commits/511aa4e2052632f16523779604bf484a1e4ab4a8#diff-fa602a8a75dc9dcc92261bac5f533c2a85e34fcceaff63b3a3a81d9acde2fc52R112). We moved those tests to a new file, so that custom allclose tolerance had to move, but also I found that I had to increase the tolerance a fair bit. This might be due to all the transposing exacerbating floating point rounding differences (though that doesn't sound plausible when I type it out), or simply the seed changed and this new seed is unlucky. Here's the failure that I was getting:

```
>       assert allclose(sim.data[p][0], np.ravel(truth))
E       assert False
E        +      where <function ravel at 0x7f18d0880430> = np.ravel

../nengo/nengo/tests/test_transforms_conv.py:151: AssertionError
---------------------------------- Captured stdout call ----------------------------------
allclose first 1 failures:
  (144,): 0.3311653 0.331148611031138
================== relative root mean squared error for allclose checks ==================
mean relative RMSE: 0.00000 +/- 0.0000 (std)
```

So it's a negligible difference in one array element. Probably just unlucky!
